### DISCUSSION TITLE
Root devices, can not get root permissions.

### DISCUSSION THREAD

**Discussion by zll123dxd** [Mygod/VPNHotspot#314] [Q&A]
Hi , I am a wifi developer.

I hope to build this feature (Vpn hotspot) in my Android 12/10 Devices.
So I hope to refer to the ip route/rule/iptables of this software.
When i install this app ,always prompt root missing.
I investigated the log . 

01-06 16:46:59.780 22630 25286 W RoutingManager: Caused by: be.mygod.librootkotlinx.NoShellException: Root missing
01-06 16:46:59.780 22630 25286 W RoutingManager: Caused by: java.io.IOException: error=13, Permission denied

root cause :  RootServer.kt ::
private fun doInit(context: Context, niceName: String) {
        val (reader, writer) = try {
            process = ProcessBuilder("su").start()
            val token1 = UUID.randomUUID().toString()
            val writer = DataOutputStream(process.outputStream.buffered())
            writer.writeBytes("echo $token1\n")
            writer.flush()
            val reader = process.inputStream.bufferedReader()
            reader.lookForToken(token1)
            Logger.me.d("Root shell initialized")
            reader to writer
        } catch (e: Exception) {
            throw NoShellException(e)   << throw this exception
        }
}

I tried it with an Android 12/10 device and the result was the same.
Devices is can adb root ,  version is userdebug.
In addition, I tried to install it as a system app , and use chmod 777 to give permission to apk.
Is this a bug? I installed it incorrectly.
Please confirm

**Comment by zll123dxd** [comment]
I'm Sorry , Devices is not root .
I'm use adb root is successful. But devices installed app can not get root permission.
Can you provide a method for constructing iptables/ ip rule/ip route, I hope to build this function in Android 10/12.

Thank you so much

**Reply by Mygod** [reply]
Take a look at `Routing.kt` but it is recommended to use this app instead to get dynamic rules updates.
